docs/source/en/api/pipelines/unidiffuser.mdx

```diff
@@ -0,0 +1,99 @@
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+-->
+
+# UniDiffuser
+
+The UniDiffuser model was proposed in [One Transformer Fits All Distributions in Multi-Modal Diffusion at Scale](https://arxiv.org/abs/2303.06555) by Fan Bao, Shen Nie, Kaiwen Xue, Chongxuan Li, Shi Pu, Yaole Wang, Gang Yue, Yue Cao, Hang Su, Jun Zhu.
+
+The abstract of the paper is the following:
+
+*This paper proposes a unified diffusion framework (dubbed UniDiffuser) to fit all distributions relevant to a set of multi-modal data in one model. Our key insight is -- learning diffusion models for marginal, conditional, and joint distributions can be unified as predicting the noise in the perturbed data, where the perturbation levels (i.e. timesteps) can be different for different modalities. Inspired by the unified view, UniDiffuser learns all distributions simultaneously with a minimal modification to the original diffusion model -- perturbs data in all modalities instead of a single modality, inputs individual timesteps in different modalities, and predicts the noise of all modalities instead of a single modality. UniDiffuser is parameterized by a transformer for diffusion models to handle input types of different modalities. Implemented on large-scale paired image-text data, UniDiffuser is able to perform image, text, text-to-image, image-to-text, and image-text pair generation by setting proper timesteps without additional overhead. In particular, UniDiffuser is able to produce perceptually realistic samples in all tasks and its quantitative results (e.g., the FID and CLIP score) are not only superior to existing general-purpose models but also comparable to the bespoken models (e.g., Stable Diffusion and DALL-E 2) in representative tasks (e.g., text-to-image generation).*
+
+Resources:
+
+* [Paper](https://arxiv.org/abs/2303.06555).
+* [Original Code](https://github.com/thu-ml/unidiffuser).
+
+Available Checkpoints are:
+- *UniDiffuser-v0 (512x512 resolution)* [dg845/unidiffuser-diffusers-v0](https://huggingface.co/dg845/unidiffuser-diffusers-v0)
+- *UniDiffuser-v1 (512x512 resolution)* [dg845/unidiffuser-diffusers-v1](https://huggingface.co/dg845/unidiffuser-diffusers)
+
+## Available Pipelines:
+
+| Pipeline | Tasks | Demo
+|---|---|:---:|
+| [UniDiffuserPipeline](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pipeline_unidiffuser.py) | *Joint Image-Text Gen*, *Text-to-Image*, *Image-to-Text*, *Image Gen*, *Text Gen*, *Image Variation*, *Text Variation* |  |
+
+## Usage Example
+
+```python
+import requests
+import torch
+from PIL import Image
+from io import BytesIO
+
+from diffusers import UniDiffuserPipeline
+
+device = "cuda"
+model_id_or_path = "dg845/unidiffuser-diffusers"
+pipe = UniDiffuserPipeline.from_pretrained(model_id_or_path)
```

Comment:
For my understanding, does this currently support (as in terms of quality):

```python
import torch

pipe = UniDiffuserPipeline.from_pretrained(model_id_or_path, torch_dtype=torch.float16)
```

This is almost a defacto standard for performing inference in `diffusers`.

Reply:
Currently there are no errors with mixed precision inference, but there may be some issues with quality (which may or may not be related to mixed precision); see discussion about generation quality.

Reply:
Hmm. Interesting.

Does using mixed-precision in the original implementation also suffer from this issue? Also, does it happen for both the checkpoints? 

Maybe there needs to be intermediate upcasting (softmax ops, for example) to prevent numerical overflows? For pipelines like [`StableUnCLIPPipeline`](https://huggingface.co/docs/diffusers/api/pipelines/stable_unclip) we use this. See: 

https://huggingface.co/stabilityai/stable-diffusion-2-1-unclip-small/blob/main/unet/config.json

Reply:
TL;DR: image-to-text samples may be messed up (other modes are probably fine?), which is true for both the v0 and v1 checkpoints. I think this is separate from numerical stability issues but haven't been able to test the full checkpoint in full precision. Using something like `upcast_attention=True` doesn't seem to solve the issue.

----

To be more precise about the issue, when doing image-to-text generation, the output text seems to be messed up. In particular, for the `diffusers` version of the UniDiffuser-v1 model, if I generate samples (with `num_inference_steps=20` and `guidance_scale=8.0`) in the different modes that produce text samples:

```bash
# Unconditional ("joint") image and text generation
A bowl of pasta with tomatoes and greens
```
![unidiffuser_sample_joint_image](https://github.com/huggingface/diffusers/assets/58458699/9f581b22-00ef-4c39-9909-c06ca49aa944)

Both the generated text and image seem reasonable, and they also match each other.

Similarly, in text-only generation mode, we get

```bash
# Text-only ("marginal text") generation
A path lined with blue pines in the woods
```

But in image-to-text mode, when I use the [UniDiffuser example image](https://github.com/thu-ml/unidiffuser/blob/main/assets/space.jpg) as the input image, I get the following output:

```bash
# Image-to-text generation
...: 'Inclothes: 'Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes
```

I've generated multiple samples (all in mixed precision) and they all end up like the above sample. Increasing `num_inference_steps` neither makes the text samples more fluent nor more accurate in describing the image.

Unfortunately, since I'm not able to generate samples from the full checkpoint using the original implementation on my computer, I'm not sure whether this is normal for the model or not. For example, when we ran the original implementation to generate expected output for the image-to-text slow test (which uses the same image and `num_inference_steps=3`), the text output is

```
T CL CL CL TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK T
```

so these types of outputs might not be unusual. On the other hand, the original model definitely generates good samples at least sometimes (e.g. the image-to-text samples from the [original paper](https://arxiv.org/pdf/2303.06555.pdf)).

I think this is distinct from the numerical stability issue mentioned in a previous comment (https://github.com/huggingface/diffusers/pull/2963#issuecomment-1547028401), although presumably the two issues interact with each other. I've tested a little with `upcast_attention=True` and it doesn't seem to solve the image-to-text bad samples issue. (I haven't tested whether it will solve the numerical issues in `torch 1.x` yet.)

> Does using mixed-precision in the original implementation also suffer from this issue?

My understanding is that the original implementation loads the models in full precision but then [runs the sampling loop in half precision](https://github.com/thu-ml/unidiffuser/blob/main/sample_multi_v1.py#L302). I haven't tested modifying this behavior yet. As for whether it has the image-to-text sample issue, see discussion above.

> Also, does it happen for both the checkpoints?

The image-to-text sample issue happens for both the UniDiffuser-v0 and UniDiffuser-v1 checkpoints.

Reply:
Thanks for your detailed answer. If image-to-text generation issue happens with the original implementation code as well, do you think it could make sense to check with the authors? 

Also, for testing purposes, I have gone ahead and added you to this Space (that access to an A10G): https://huggingface.co/spaces/unidiffuser-testing/unidiffuser-testing.

Hopefully, that will help you to run the tests on the full checkpoint?

Reply:
I think I have debugged the error, turns out when encoding the image using the VAE I forgot to scale the latents by the VAE's scaling factor 😅.

Reply:
Damn it that little bugger! Can be very brutal when missed!